File: Ssafy/src/Graph/DFS.java
package Graph;

import Util.UtilClass;

import java.util.ArrayList;
import java.util.HashMap;
import java.util.Scanner;

public class DFS {
    static UtilClass util = new UtilClass();
    static Scanner sc = new Scanner(System.in);

    public static void main(String[] args) {
        HashMap<String, ArrayList<String>> graph = new HashMap<>();
        util.inputHashMap(graph);

        util.printStringList(dfs(graph, "A"));
    }

    public static ArrayList<String> dfs(HashMap<String, ArrayList<String>> graph, String startNode) {
        ArrayList<String> visited = new ArrayList<>();
        ArrayList<String> needVisited = new ArrayList<>();

        needVisited.add(startNode);

        while (needVisited.size() > 0) {
            String node = needVisited.remove(needVisited.size() - 1);

            if (!visited.contains(node)) {
                visited.add(node);
                needVisited.addAll(graph.get(node));
            }
        }

        return visited;
    }
}

File: Theory/src/DFS_BFS/DFS.kt
import java.util.*

var n: Int = 0
var m: Int = 0
var graph = Array(101) { Array(101) { 0 } }
var visit = Array(101) { 0 }

fun main(args: Array<String>) = with(Scanner(System.`in`)) {
    var a: Int
    var b: Int

    n = nextInt()
    m = nextInt()

    for (i in 0 until m) {
        a = nextInt()
        b = nextInt()
        graph[a][b] = 1
        graph[b][a] = 1
    }

    dfs(1)

    println(visit.filter { it == 1 }.size - 1)
    println(visit.toList())
}

fun dfs(x: Int) {
    visit[x] = 1
    for (i in 1..n) {
        if (visit[i] == 1 || graph[x][i] == 0) continue
        dfs(i)
    }
}
